components/logins/src/update_plan.rs

```diff
@@ -219,7 +219,10 @@ impl UpdatePlan {
                  hostname            = :hostname,
                  username            = :username,
                  sync_status         = {changed}
-             WHERE guid = :guid",
+             WHERE guid = :guid
```

Comment:
The inductive argument is:
1. If the database has no duplicate hostnames, then `guid` → `hostname` is 1:1.
2. Every update preserves the injection, since it cannot change a mapping.

However, after looking at this, I am curious if it is a valid operation to change any of these fields for a particular `guid`. If so, then this needs to be more sophisticated. In a single query it would be:

```
WHERE guid = :guid
AND NOT EXISTS (
  SELECT guid FROM loginsL
  WHERE guid != :guid
  AND (
    hostname = :hostname
    OR httpRealm = :http_realm
    OR formSubmitURL = :form_submit_url
   ))
```

I believe that is correct: the result set will be empty when there is no other row that has the name hostname, http realm, or form url. This does result in 2 queries for each update. I am not sure whether there are performance constraints/concerns here (I don't know how efficient/inefficient the db engine is, nor what the data set size is, though I imagine it's not that big?).

Can you get confirmation that:
1. It is valid to change hostname, http realm, or url for a given `guid` and
2. A subquery is not too expensive to be doing.

Thanks!

Reply:
Didn't you say that `username` is also part of what makes this record unique? If that's the case then you'll need to do the same with username here.

Also verify if a subquery is not too expensive.

I think there should also be a test for this. Is this exercised through other tests of higher-level components? You could probably add a test that exercises this in components/logins/src/db.rs. See [here](https://github.com/mozilla/application-services/blob/d4d4f6609b99766bb082a6ddc327cefc6735fee7/components/logins/src/db.rs#L1363), for example.

Reply:
Yes you're right, I just added a check for the username, too.

I asked in slack about the subquery, so I'll tell you what they say.

In the file you linked, is test_check_valid_with_no_dupes() already testing that duplicates are rejected? It tests a unique login and a duplicate login. Should I add other logins that are supposed to be updates to the unique login, but not dupes?

Reply:
I don't think that test is really exercising your code. You want to add a few logins and then execute an update plan with the following permutations:

- update would create equality w/ other records for some fields, but not all, and therefore should succeed.
- update would create equality w/ other records for all fields, and therefore should fail.
- update doesn't overlap and therefore should succeed.
- etc.

You can see tests in the same file that use update plans to do similar, but not the same thing. For example: https://github.com/mozilla/application-services/blob/b6fccf849906ea764b41fd1eeea6a7ffd4388d0a/components/logins/src/db.rs#L1078

Reply:
So I got an answer to the questions you wanted me to ask. 

1. "It would be OK to change those values, so long as the change didn't create a dupe." 
2. "I think it would be possible to do a query to check for dupes, we just haven't yet, and there's also some small concern about normalizing the data (ie, we previously stored data in the wrong format - eg, a full URL where we should have just stored the origin), and that might be tricky to resolve in a sub-query)"

So do you think the query we did to check for dupes will work knowing this?

Reply:
Yep, sounds like this is a fine approach. You can go ahead and add tests now. :)